### ISSUE TITLE
How to finetune the multi-step prediction？  I get the stable frames over time step.

### ISSUE THREAD

**Issue by flyingmoth** [coxlab/prednet#35]
I first train the model in t+1 prediction. Then finetune the model with "extrap_start_time=10" and nt=15.
But the result frames are stable from t=11 to 15.

**Comment by bill-lotter** [comment]
Hi, I'm not sure why that would happen, but we've recently discovered an issue in Keras 2 with TimeDistributed weight initialization (https://github.com/keras-team/keras/issues/8895). I've since adjusted the kitti_train.py code for this (https://github.com/coxlab/prednet/commit/9f6482e6e6038e02aa3dbd878bae23c7b3a5a28a). Try running with the new code and see if you have similar issues - thanks!

**Comment by ReDeiPirati** [comment]
Hi @flyingmoth 
May I ask for how long you have trained both for t+1 prediction and for fine-tuning?

Hi @bill-lotter,
In the paper is not specified for how long you have trained for fine-tuning the model on multi-step prediction after the t+1 training... May I ask you to provide some info, please? Thanks again for the code and the great paper!

**Comment by bill-lotter** [comment]
Hi @ReDeiPirati,

For the fine-tuning, I trained it with the same parameters (150 epochs, decreasing learning rate at 75) and yes, the only changes are nt=15, extrap_start_time=10 and starting from the t+1 trained weights. Thanks for the support!

**Comment by abe-** [comment]
Hi,

I'm finding the same behaviour as @flyingmoth, with the last version of the code. I've trained initially a model for t+1 with nt = 10 and then fine tuned it with extrap_start_time=10 and nt=15. 

(To do this fine tuning, I've just executed again the train script with the new nt and extrap_start_time parameters, and left untouched the initially generated t+1-model files in their location). 

The resulting predicted frames 11 to 15 appear however as the same unchanged image.

I'm using keras 2.0.6 and tensorflow 1.2.1

Thanks in advance, and for this great work.

**Comment by ReDeiPirati** [comment]
Unfortunately the same has happened to me: tested on Keras 2.0.8 and TF 1.3/1.4.

The t+1 prediction works fine.

**Comment by bill-lotter** [comment]
Hi @flyingmoth @ReDeiPirati 

I figured out what the issue is. I had forgotten that if you train in extrapolation mode, the "error" representation in the model doesn't correspond to the ground truth error representation anymore because the "actual" is just the prediction from the previous timestep, thus if the loss is computed from the errors, it doesn't make sense. The easiest thing to do is to just directly compute the MAE between the frame predictions and ground truth sequence, and use this as the loss, which is otherwise equivalent to the "error" loss when running in normal t+1 mode with the L_0 model. This is what I had done for the results in the paper and for the provided weights, I had just forgotten about it. Sorry about that - my apologies! I added the script `kitti_extrap_finetune.py` to show how this is done, should have just included it to begin with.

**Comment by abe-** [comment]
Hi @bill-lotter,

I've just confirmed that the finetuned multi-step prediction works now in my setup.

Thanks!
